### PR TITLE
Make unused variables with inferred type including `error` result in a compilation error

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title, when the binding pattern is a capture binding pattern that binds a (potential) error.

The following will now result in an error.
```ballerina
import ballerina/io;

public function main() {
    var res = update(11); // error: unused variable 'res' with inferred type including error
    io:println("Done");
}

function update(int id) returns error? {
    if id < 10 {
        // Update.
        return;
    }

    return error("Invalid id: " + id.toString());
}
```

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/28600

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 1]
```diff
@@ -278,11 +280,18 @@ public static DataflowAnalyzer getInstance(CompilerContext context) {
...
         this.functionToDependency = new HashMap<>();
         this.dlog.setCurrentPackageId(pkgNode.packageID);
         SymbolEnv pkgEnv = this.symTable.pkgEnvMap.get(pkgNode.symbol);
         analyzeNode(pkgNode, pkgEnv);
+
+        for (Map.Entry<BSymbol, Location> entry : this.unusedErrorVarsDeclaredWithVar.entrySet()) {
+            this.dlog.error(entry.getValue(), DiagnosticErrorCode.UNUSED_VARIABLE_WITH_INFERRED_TYPE_INCLUDING_ERROR,
+                            entry.getKey().name);
+        }
+
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:294]: Can we move this to `visit(BLangPackage pkgNode)`

**Comment by MaryamZi** [comment]
Marking as WIP since we need to check with the standard library team re: downstream impact before merging.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java (hunk 1 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Fixed in 415c3aa.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 2]
```diff
@@ -952,6 +961,7 @@ public void visit(BLangRecordLiteral recordLiteral) {
 
     @Override
     public void visit(BLangSimpleVarRef varRefExpr) {
+        unusedErrorVarsDeclaredWithVar.remove(varRefExpr.symbol);
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:964]: What is the reason for removing here?

**Comment by KRVPerera** [comment]
Isn't this an issue.
```
int|error a = 4;

public function main() {
    if (a is int) {
        var s = a; // unused variable 's' with inferred type including error
    }
}

```

**Review by KRVPerera — APPROVED** [review]
LGTM.

**Pull request merged**